app/routes/surveys/components/Question.js

```diff
@@ -0,0 +1,237 @@
+// @flow
+
+import React, { Component } from 'react';
+import Option from './Option';
+import { TextInput, TextArea, CheckBox } from 'app/components/Form';
+import Dropdown from 'app/components/Dropdown';
+import styles from './surveys.css';
+import Icon from 'app/components/Icon';
+import { Link } from 'react-router';
+import { ConfirmModalWithParent } from 'app/components/Modal/ConfirmModal';
+
+type Props = {
+  updateQuestion: Object => void,
+  deleteQuestion: number => Promise<*>,
+  question: Object,
+  nr: number
+};
+
+type State = {
+  shouldUpdate: boolean,
+  options: Array<Object>,
+  pickerOpen: boolean
+};
+
+type QuestionTypeSelectorProps = {
+  questionType: number
+};
+
+function QuestionTypeSelector({ questionType }: QuestionTypeSelectorProps) {
+  const questionTypeIntToString = {
+    '1': 'Multiple Choice',
+    '2': 'Sjekkboks',
+    '3': 'Fritekst'
+  };
+  const questionTypeIntToIcon = {
+    '1': 'radio-button-on',
+    '2': 'checkbox',
+    '3': 'more'
+  };
+  return (
+    <Link className={styles.questionTypeSelector}>
+      <span style={{ width: '170px', display: 'inline-block' }}>
+        <Icon
+          name={questionTypeIntToIcon[String(questionType)]}
+          style={{ marginRight: '15px' }}
+        />
+        {questionTypeIntToString[String(questionType)]}
+      </span>
+      <Icon name="arrow-down" size={24} className={styles.typeIcon} />
+    </Link>
+  );
+}
+
+type DropdownItemsProps = {
+  setQuestionType: number => void
+};
+
+function DropdownItems({ setQuestionType }: DropdownItemsProps) {
+  return (
+    <Dropdown.List>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(1)}>
+          <strong>Multiple Choice</strong>
+          <Icon name="radio-button-on" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(2)}>
+          Sjekkboks
+          <Icon name="checkbox" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(3)}>
+          Fritekst
+          <Icon name="more" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+    </Dropdown.List>
+  );
+}
+
+class Question extends Component<Props, State> {
+  state = {
+    shouldUpdate: false,
+    options: [{ nr: 0, optionText: '' }],
+    pickerOpen: false
+  };
+
+  componentDidMount() {
+    const options = this.props.question.options
+      .map((option, i) => ({ ...option, nr: i }))
+      .concat({
+        nr: this.props.question.options.length,
+        optionText: ''
+      });
+    this.setState({ options });
+  }
+
+  componentDidUpdate() {
+    if (this.state.shouldUpdate) {
```

Comment:
this `shouldUpdate` thing seems fishy - could you try explaining what it does?

Reply:
Not completely sure I understand, but maybe you were calling `setState` after mutating `options`? `setState` is sort of like Redux in that regard: you need a new reference, i.e.:

```js
// Bad:
this.state.options.push(newOption);
this.setState({ options: this.state });

// Good:
const options = [...this.state.options, newOption];
this.setState({ options });
```

I saw that you used `.slice()` to create a copy further down, but this only creates a shallow copy of the array - the option objects inside the array are still the same, so if you mutate any of those it won't matter (not sure if you do though).

https://medium.com/pro-react/a-brief-talk-about-immutability-and-react-s-helpers-70919ab8ae7c

If that's not the case we can discuss it later today, at the meeting.